New bonus guide: LNDg

VajraOfIndra opened the pull request:
#### What

LNDg is a lite GUI web interface to help you manually manage your node and automate operations such as rebalancing, fee adjustemnts and channel node opening.

Has been requested by the community: e.g. https://t.me/raspibolt/12058

#### How

A new bonus guide in the +Lightning section

#### Scope

- [ ] significant change to core configuration
- [x] independent bonus guide
- [ ] simple bug fix

Fixes #1125

#### Test & maintenance

How can the changes be tested? Is there ongoing maintenance effort? If this is a bonus guide: are you willing to update it from time to time?

#### Animated GIF (optional)

![hammock](https://media.giphy.com/media/piO6cmvxIK0A05MNkY/giphy-downsized.gif)

twofaktor commented:
Where is the `Remote Tor connection`?

VajraOfIndra commented:
> Where is the `Remote access over Tor` section? Is not necessary?

It's not necessary but would be nice to have for sure. You could make a PR in my repo if you want, or else someone could add it later once it's merged. I let you manage that as i wont have time to add the Tor sectio myself.

martinatime commented on guide/bonus/lightning/lndg.md:
````diff
@@ -0,0 +1,662 @@
+  $ cd lndg
+  ```
+
+* As user “lndg”, create a bash file to call `rebalance.py` and paste the following lines. Save and exit.
````
I think this is the wrong python script for this section

VajraOfIndra replied:
fixed

martinatime commented on guide/bonus/lightning/lndg.md:
````diff
@@ -0,0 +1,662 @@
+  $ exit
+  ```
+
+* Create a service file for the `rebalancer.sh` script
````
Wrong shell script for this section

VajraOfIndra replied:
Fixed (I think I got them all, PTAL)

martinatime commented on guide/bonus/lightning/lndg.md:
```diff
@@ -0,0 +1,662 @@
+
+---
+
+## For the future: LNBits update
```
LNBits? Do you mean LNDg?

VajraOfIndra replied:
Fixed! Thanks

martinatime commented:
I think this guide is probably in mostly good shape. I left some comments where I think there was some copy/paste errors.

VajraOfIndra commented:
> I think this guide is probably in mostly good shape. I left some comments where I think there was some copy/paste errors.

I have some issue with the uwsgi/nginx section. It comes a pull request I merged without testing myself at the time. I think I might remove that section for the time being. We could always add it later via a tested PR once it works. Same with remote connection through Tor.` Wdyat?

blckbx commented on guide/bonus/lightning/lndg.md:
```diff
@@ -0,0 +1,783 @@
+  [Unit]
+  Description=LNDg uWSGI app
+  After=lnd.service
+  PartsOf=lndg.service
```
```suggestion
  PartOf=lnd.service
```

VajraOfIndra commented:
I reviewd the guide and fixed a few major issues:
- first start URL is clearnet http (not https)
- removed gloabl apt install uwsgi (is installed via Python pip)
- Removed outdated StandardError of service file
- made the backup services (jobs, rebalancer etc) dependent on lnd.service (suing After and PartOf). We could make them dependent on uwsgi.service maybe?
The guide works on my node.

VajraOfIndra commented:
I simplified the guide by making the backend services and timers dependent on `uwsgi.service`, which is itself dependent on `lnd.service`.

Like this, to stop or start LNDg and the backend services (e.g., when updating), we'll just need to start and stop `uwsgi.service` (all the other timers and services will start and stop automatically). And when we start and stop LND, LNDg/`uwsgi.service` will start and stop as well automatically.

|Service|`After`|`PartOf`|
|-------|------|-------|
|`uwsgi.service`|`lnd.service`|`lnd.service`|
|`jobs-lndg.service`|||
|`jobs-lndg.timer`|`uwsgi.service`|`uwsgi.service`|
|`rebalancer-lndg.service`|||
|`rebalancer-lndg.timer`|`uwsgi.service`|`uwsgi.service`|
|`htlc-stream-lndg.service`|`uwsgi.service`|`uwsgi.service`|

blckbx commented on guide/bonus/lightning/lndg.md:
````diff
@@ -0,0 +1,796 @@
+  ```sh
+  $ sudo adduser --disabled-password --gecos "" lndg
+  $ sudo adduser lndg lnd
+  $ sudo adduser lndg www-data
````
In the end, I also got `502 gateway timeout` as mentioned in #1125. Solution was to add `www-data` to `lndg` group:
```sh
$ sudo adduser www-data lndg
```

VajraOfIndra replied:
I added that step

blckbx commented on guide/bonus/lightning/lndg.md:
````diff
@@ -0,0 +1,796 @@
+  WantedBy=sockets.target
+  ```
+
+* Create a nginx configuration file for the LNDg website with a server listening on port 4890
````
The listening port is set to 8889 in config below.

blckbx commented on guide/bonus/lightning/lndg.md:
```diff
@@ -0,0 +1,796 @@
+
+### Blockchain explorer
+
+To preserve privacy it is better that you use your own self-hosted blockhain explorer (e.g. the BTC RPC Explorer. 
```
```suggestion
To preserve privacy it is better that you use your own self-hosted blockhain explorer (e.g. the BTC RPC Explorer).
```

blckbx commented on guide/bonus/lightning/lndg.md:
```diff
@@ -0,0 +1,796 @@
+
+To preserve privacy it is better that you use your own self-hosted blockhain explorer (e.g. the BTC RPC Explorer. 
+
+* Open your LNDg website at https://raspibolt.local:8889 (replace raspibolt.local by your nod's IP address if necessary)
```
```suggestion
* Open your LNDg website at https://raspibolt.local:8889 (replace raspibolt.local by your node's IP address if necessary)
```

blckbx reviewed (commented):
some nits, one bugfix, rest works like a charm! 🥳

blckbx reviewed (approved):
tACK (amd64)

VajraOfIndra commented:
> tACK (amd64)

Thanks for the review. Is it ready to merge you think?

Btw, just made a little [change ](https://github.com/raspibolt/raspibolt/pull/1070/commits/d6d3d60a7e7e906a9b161e3c8fa9ce2bcba4b172)to make the backend service and timers start when uwsgi starts (as PartOf does not take care of that). 

To test it:
* Add the new line
* Register the change
  `$ sudo systemctl daemon-reload`

* Stop LNDg
  `$ sudo systemctl stop uwsgi.service`

* Check that `jobs-lndg.timer` has stopped (triggered by the `PartOf` option in `jobs-lndg.timer`)
  ```
  $ sudo systemctl status jobs-lndg.timer
  > [...]
  >  Active: inactive (dead)
  ```

* Starts LNDg
  `$ sudo systemctl start uwsgi.service`

* Check that `jobs-lndg.timer` started automatically (trigerred by the `Wants` option in `uwsgi.service`)
  ```
  $ sudo systemctl status jobs-lndg.timer
  > [...]
  >  Active: active (running)
  ```

blckbx commented:
> Thanks for the review. Is it ready to merge you think?

Regarding the last change:  I already updated my local installation and it works perfectly fine.

I think the PR is in very good shape, it's probably a good idea to wait for feedback on your TG question that could strip down the guide quite a bit: https://t.me/c/1442363505/33435

Other than that: another thing mentioned in LNDg group and also done in RaspiBlitz' PR: they moved `/home/lndg/lndg/data` directory to `/data/lndg/data` and symlinked it. wdyt about that?

VajraOfIndra commented:
> Regarding the last change: I already updated my local installation and it works perfectly fine.

Good to hear!

> I think the PR is in very good shape, it's probably a good idea to wait for feedback on your TG question that could strip down the guide quite a bit: https://t.me/c/1442363505/33435

Good idea, will wait to hear from NGU

> Other than that: another thing mentioned in LNDg group and also done in RaspiBlitz' PR: they moved `/home/lndg/lndg/data` directory to `/data/lndg/data` and symlinked it. wdyt about that?

This seems to make sense indeed. Nice catch.
What goes in the LNDg data folder? some sort of database?

blckbx commented:
> This seems to make sense indeed. Nice catch. What goes in the LNDg data folder? some sort of database?

Yeah, settings, stats, everything worth being stored.

VajraOfIndra commented:
> Other than that: another thing mentioned in LNDg group and also done in RaspiBlitz' PR: they moved `/home/lndg/lndg/data` directory to `/data/lndg/data` and symlinked it. wdyt about that?

I added a section to do this:
* used rsync instead of cp to avoid corruption of DB
* stored the DB in `/data/lndg` rather than `/data/lndg/data` as extra directory not necessary
* new section is done after the step that removes the password file as this file is in the same data directory as the DB

PTAL :)

blckbx commented on guide/bonus/lightning/lndg.md:
````diff
@@ -144,6 +144,37 @@ If you didn't save the password, you can get it again with: `nano /home/lndg/lnd
+* Delete the old database from the LNDg runtime directory and instead create a symbolic link 
+
+  ```sh
+  $ sudo rm /home/lndg/lndg/data/db.sqlite
````
```suggestion
  $ sudo rm /home/lndg/lndg/data/db.sqlite3
```

blckbx reviewed (commented):
bad filename

blckbx commented:
tACK²
